src/main/scala/com/banzaicloud/spark/metrics/SparkCollectorDecorators.scala

```diff
@@ -0,0 +1,119 @@
+package com.banzaicloud.spark.metrics
+
+import java.util
+
+import com.banzaicloud.spark.metrics.PushTimestampDecorator.PushTimestampProvider
+import com.codahale.metrics.MetricRegistry
+import io.prometheus.client.Collector
+import io.prometheus.client.Collector.MetricFamilySamples
+import io.prometheus.client.Collector.MetricFamilySamples.Sample
+import io.prometheus.client.dropwizard.DropwizardExports
+import io.prometheus.jmx.JmxCollector
+import org.apache.spark.internal.Logging
+
+import scala.collection.JavaConverters._
+import scala.util.matching.Regex
+
+object NameDecorator {
+  case class Replace(Regex: Regex, replacement: String)
+}
+
+trait NameDecorator extends CollectorDecorator {
+  val metricsNameReplace: Option[NameDecorator.Replace]
+
+  protected override def familyName(familySamples: Collector.MetricFamilySamples): String = {
+    replaceName(super.familyName(familySamples))
+  }
+  protected override def sampleName(sample: Sample): String = {
+    replaceName(super.sampleName(sample))
+  }
+
+  private def replaceName(name: String) = {
+    metricsNameReplace.map {
+      case NameDecorator.Replace(regex, replacement) => regex.replaceAllIn(name, replacement)
+    }.getOrElse(name)
+  }
+}
+
+trait LabelsDecorator extends CollectorDecorator {
+    val extraLabels: Map[String, String]
+
+    private val labelNames = extraLabels.keys.toList.asJava
+    private val labelValues = extraLabels.values.toList.asJava
+
+    override protected def sampleLabelNames(sample: Sample): util.List[String] = {
+      mergeLists(super.sampleLabelNames(sample), labelNames)
+    }
+
+    override protected def sampleLabelValues(sample: Sample): util.List[String] = {
+      mergeLists(super.sampleLabelValues(sample), labelValues)
+    }
+
+    private def mergeLists(list1: util.List[String], list2: util.List[String]): util.List[String] = {
+      val newList = new util.ArrayList[String](list1)
+      newList.addAll(list2)
+      newList
+    }
+}
+
+object PushTimestampDecorator {
+  case class PushTimestampProvider(getTimestamp: () => Long = System.currentTimeMillis) extends AnyVal
+}
+trait PushTimestampDecorator extends CollectorDecorator {
+  val maybeTimestampProvider: Option[PushTimestampProvider]
+
+  override protected def sampleTimestamp(sample: Sample): java.lang.Long = {
+    maybeTimestampProvider match {
+      case Some(t) => t.getTimestamp()
+      case None => sample.timestampMs
+    }
+  }
+}
+
+trait ConstantHelpDecorator extends CollectorDecorator {
+  val constatntHelp: String
+
+  override protected def helpMessage(familySamples: MetricFamilySamples): String = {
+    constatntHelp
+  }
+}
+
+trait DeduplicatorDecorator extends CollectorDecorator with Logging {
+
+  override def collect(): util.List[Collector.MetricFamilySamples] = {
```

Comment:
It's not enough to deduplicate metrics at collector level as there might be multiple collectors. The deduplication must happen across all collectors which metrics all collected from before pushing to Pushgateway.

Reply:
Good point. In the latest commit I added DeduplicatedCollectorRegistry which do that.